src/haddock/libs/libplots.py

```diff
@@ -523,29 +561,317 @@ def scatter_plot_handler(capri_filename, cl_rank, png, dpi):
+        fig.update_yaxes(
+            title_text=sub_fig.layout.yaxis.title.text,
+            row=row_index,
+            col=col_index,
+            title_standoff=5,
+            automargin=True,
+            )
+        # x title only on the last row
+        if shared_xaxes == "all":
+            row_index = number_of_rows
+        fig.update_xaxes(
+            title_text=sub_fig.layout.xaxis.title.text,
+            row=row_index,
+            col=col_index,
+            title_standoff=5,
+            automargin=True,
+            )
+        legend_title_text = sub_fig.layout.legend.title.text
+    fig.update_layout(
+        title_text=plot_title,
+        legend_title_text=legend_title_text,
+        height=height * number_of_rows,
+        width=width * number_of_cols,
+        )
+    return fig
+
+
+def find_best_struct(ss_file, number_of_struct=4):
+    """
+    Find best structures.
+
+    It inspects model-cluster-ranking recorded in capri_ss.tsv file and finds
+    the best models (models with lower ranks). By default, it selects the 4 best
+    models.
+
+    Parameters
+    ----------
+    ss_file : path
+        path to capri_ss.tsv
+    number_of_struct: int
+        number of models with lower model-cluster-ranking
+
+    Returns
+    -------
+    best_struct_df : pandas DataFrame
+        DataFrame of best structures
+    """
+    dfss = read_capri_table(ss_file)
+    dfss = dfss.sort_values(by=["cluster-id", "model-cluster-ranking"])
+    # TODO need a check for "Unclustered"
+    # possible number of structures, can be different per each cluster
+    max_number_of_struct = dfss.groupby("cluster-id").count()["model-cluster-ranking"].min()
+    number_of_struct = min(number_of_struct, max_number_of_struct)
+    best_struct_df = dfss.groupby("cluster-id").head(number_of_struct).copy()
+    number_of_cluster = len(best_struct_df["cluster-id"].unique())
+    col_names = [
+        f"Nr {number + 1} best structure" for number in range(number_of_struct)
+        ] * number_of_cluster
+    best_struct_df = best_struct_df.assign(Structure=col_names)
+    best_struct_df = best_struct_df.pivot_table(
+        index=["cluster-id"],
+        columns=["Structure"],
+        values="model",
+        aggfunc=lambda x: x,
+        )
+    best_struct_df.reset_index(inplace=True)
+    best_struct_df.rename(columns={"cluster-id": "Cluster ID"}, inplace=True)
+    return best_struct_df
+
+
+# TODO links to be fixed!
+def _add_links(best_struct_df):
+    table_df = best_struct_df.copy()
+    for col_name in table_df.columns[1:]:
+        file_name = best_struct_df[col_name].apply(lambda row: Path(row).name)
+        # href empty for local files
+        # TODO fix links to Download and Visibility
+        # we can use html href syntax but html should be responsive!
+        dl_text = "Download"
+        vis_text = "Visibility"
+        table_df[col_name] = file_name + ", " + dl_text + ", " + vis_text
+    return table_df
```

Comment:
I guess that this is the function that should create all the links to the structures and to their visualization..the problem here is that in a standard run we might have hundreds (if not thousands) of clusters, and including all the links here would probably end up in a huge html report, what do you think?

Reply:
good point. I am wondering how this issue was handled in haddock 2.4. Here, the table creates vertical scrolling. I committed the `report.html` generated for `-m 5`, see [my commit](https://github.com/haddocking/haddock3/pull/611/commits/996d16dae256bab94537dce146c037af12247efe). If the file size of the report is a concern, we need to think about exporting it to another format perhaps a csv file!